### PR TITLE
feat: migrate to pyproject.toml

### PR DESCRIPTION
**Pull request by rinarakaki**
Tracking issue https://github.com/googleapis/google-cloud-python/issues/13171

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+bigquery_v2 = [
+  "proto-plus >= 1.22.3, < 2.0.0dev",
+  "protobuf >= 3.20.2, < 6.0.0dev, != 4.21.0, != 4.21.1, != 4.21.2, != 4.21.3, != 4.21.4, != 4.21.5", # For the legacy proto-based types.
+]
+all = [
+  "google-cloud-bigquery[bqstorage,pandas,ipywidgets,geopandas,ipython,tqdm,opentelemetry,bigquery_v2]",
+]
+
+[tool.setuptools.dynamic]
+version = { attr = "google.cloud.bigquery.__version__" }
```
Comment (chalmerlowe) [inline pyproject.toml:99]:
```suggestion
[tool.setuptools.dynamic]
version = { attr = "google.cloud.bigquery.version.__version__" }
```

**Review by chalmerlowe — APPROVED** [review]
LGTM

**Comment by chalmerlowe** [comment]
@rinarakaki 

Thank you for your assistance with converting this to pyproject.toml.

I am curious... you went through and offered up conversion PRs for a large number of google projects. What inspired this?

**Comment by rinarakaki** [comment]
@chalmerlowe 
Thanks for reviewing this PR!

For me, it was more about how to motivate/inspire the project maintainers to migrate to pyproject.toml, because this kind of problem (build systems) can only be solved in coordination with other projects, as they tend to rely on the common internal practices, standards and tooling. I use these libraries at work and want them to keep up with the latest Python standards and trends (including typing). pyproject.toml is a good first step towards modernising these Python projects, I believe.

**Comment by chalmerlowe** [comment]
@rinarakaki 

That makes sense.
I manage: 

* python-bigquery
* python-bigquery-pandas
* python-bigquery-sqlalchemy
* python-bigquery-storage
* python-db-dtypes-pandas

and will be migrating all of these to pyproject.toml in the near future.

**Pull request merged**